Move Teleport Connect over to Vite

ryanclark opened the pull request:
This moves Connect's build & development tooling over to Vite.

In order to use Vite (which doesn't like CommonJS includes unless it's a 3rd party module), I had to swap over our JS protobuf generation to TypeScript - this is actually a nice advantage, because we can now use the protobuf types in the Teleport codebase. It also fixes a few type issues/FIXMEs in the codebase. I moved user preferences over to using the protobuf files as there was intertwined logic between Teleterm, Teleport & the gRPC representation, so I simplified it.

Sorry for the large PR. These changes were a result of changing one thing and then fixing the errors, changing another, fixing those, etc - it was hard to keep these concise and in separate, ordered commits.

### Speed improvements

Start -
Vite - 5.89 seconds
Webpack 22.7 seconds

Build -
Vite - 14.28 seconds
Vite (bytecode) - 168.26 seconds
Webpack - 92.64 seconds

### Caveats

The `oneOf` representation in `@protobuf-ts/plugin` only works if `strictNullChecks` is enabled. I tried to enable this, but I couldn't get Connect to only be checked and not Teleport.

To get around this, I added some helpers to ensure type safety. There's only 3 different oneOf values to check in the codebase, and I don't think it's too tedious to add new helpers for new values. If the protobuf representation changes, the types will break, so they will always be in sync.

### Things to test

- [x] Bundling and packaging
  - [x] `agentCleanupDaemon.js` requires `winston`, but it's not included in `node_modules`. Webpack used to bundle it into `agentCleanupDaemon.js` itself.
- [x] Passwordless login
- [x] Shell test plan
  - [x] Detecting `$HOME` doesn't seem to work, opening a new local shell tab while `DocumentCluster` is active opens `/` instead of `~`.
- [x] File transfer
- [x] [Hot module replacement](https://electron-vite.org/guide/hmr)
- [x] F6 to reload main in dev mode
  -  electron-vite supports [hot reloading of the main process](https://electron-vite.org/guide/hot-reloading), but they say "Because the timing of reloading is uncontrollable, hot reloading cannot be perfectly implemented. In practice, hot reloading is not always beneficial，so it is recommended to use the CLI option to flexibly decide whether to enable or not."
- [x] Removing unreachable code (`process.env.NODE_ENV`)
- [x] CSP matches that generated by Webpack
- `yarn dev-term` and packaged app are considered to be the same app by Electron, preventing both from running at the same time.
  - Not a blocker, we can deal with this later. Most likely a result of the dev version of the app using the same package.json (and thus package name) as the packaged prod version.
- [x] Usage events

ryanclark committed:
Move Teleport connect over to Vite

ryanclark committed:
Clean up the Electron Vite config

ryanclark committed:
Add mts to prettier/eslint

ryanclark commented:
Note - I had tried to add `strictNullChecks`, so there may be code changes to account for that, but they're not needed now. Please point out any you see and I'll revert

ryanclark committed:
Add license headers

ravicious committed:
Dedupe @babel scope

ravicious committed:
Remove console.log from mainProcess

ravicious commented on web/packages/teleterm/src/services/tshd/createClient.ts:
```diff
@@ -78,24 +86,25 @@ export function createTshdClient(
       startKey,
       limit,
     }: types.GetResourcesParams) {
-      const req = new api.GetKubesRequest()
-        .setClusterUri(clusterUri)
-        .setSearchAsRoles(searchAsRoles)
-        .setStartKey(startKey)
-        .setSearch(search)
-        .setQuery(query)
-        .setLimit(limit);
+      const req = api.GetKubesRequest.create({
+        clusterUri,
+        searchAsRoles,
+        startKey,
+        search,
+        query,
+        limit,
+      });
```
😌

ravicious committed:
Set suggestedReviewers on CreateAccessRequestRequest

ravicious commented on web/packages/teleterm/src/services/tshd/createClient.ts:
```diff
@@ -859,42 +911,42 @@ export function createTshdClient(
-                  nextKey: res.getNextKey(),
-                  resources: res
-                    .getResourcesList()
+                  nextKey: res!.nextKey,
```
Yesss, now comments from protos get generated into JSDocs. Try looking up `nextKey` here.

ravicious commented on web/packages/teleterm/src/services/tshd/createClient.ts:
```diff
@@ -903,105 +955,108 @@ export function createTshdClient(
-      }
+    ): Promise<api.UserPreferences> {
+      const userPreferences = UserPreferences.create();
+      if (params.userPreferences) {
```
That seems like a leftover from strict null checks. This check here probably doesn't do anything bad, but I guess I'd remove it just to keep the implementation as close to the previous one as possible?

ryanclark replied:
> I'd remove it just to keep the implementation as close to the previous one as possible?

Agreed, undone

ravicious commented on web/packages/teleterm/src/services/tshd/createClient.ts:
```diff
@@ -903,105 +955,108 @@ export function createTshdClient(
+        const req = api.UpdateUserPreferencesRequest.create({
+          clusterUri: params.clusterUri,
+          userPreferences: {
+            unifiedResourcePreferences:
+              userPreferences.unifiedResourcePreferences,
+            clusterPreferences: userPreferences.clusterPreferences,
```
I think we can pass `userPreferences` directly.

ryanclark replied:
Yep!

ravicious commented on build.assets/Dockerfile:
```diff
@@ -291,9 +291,8 @@ RUN helm plugin install https://github.com/vbehar/helm3-unittest && \
-ARG NODE_PROTOC_TS_VERSION  # eg, "5.0.1"
-RUN npm install --global "grpc-tools@$NODE_GRPC_TOOLS_VERSION" "grpc_tools_node_protoc_ts@$NODE_PROTOC_TS_VERSION"
+ARG PROTOBUF_TS_PLUGIN_VERSION  # eg, "2.9.3"
+RUN npm install --global "@protobuf-ts/plugin@$PROTOBUF_TS_PLUGIN_VERSION"
```
I'm not sure if we won't have to do this in three steps, where we first add `@protobuf-ts/plugin` to the Dockerfile, then add code that uses it and then remove `grpc-tools` and `grpc_tools_node_protoc_ts`.

I'm not up to date how the GitHub Actions work nowadays, but with GCP I remember issues with the fact that the builders were using cached/published version of the Dockerfile and not the one from the branch.

The lint action failed on this:

```
buf generate --template=buf-ts.gen.yaml --path=proto/prehog/ --path=proto/teleport/lib/teleterm/ --path=api/proto/teleport/userpreferences/ --path=api/proto/teleport/accesslist/ --path=api/proto/teleport/trait/ --path=api/proto/teleport/header/
Failure: plugin typescript: exec: "/usr/local/lib/node_modules/@protobuf-ts/plugin/bin/protoc-gen-ts": stat /usr/local/lib/node_modules/@protobuf-ts/plugin/bin/protoc-gen-ts: no such file or directory
```

https://github.com/gravitational/teleport/actions/runs/7585240077/job/20660640230?pr=36921#step:15:17

But I'm not 100% sure if it was caused by what I'm talking about.

ryanclark replied:
Yeah I think you're correct. Let me split out this part into a new PR

ryanclark replied:
https://github.com/gravitational/teleport/pull/37009

ravicious reviewed (commented):
I went through the tshd client. I'm yet to build the app, but I definitely want to look into how the final packaged app looks like structure-wise. I can also do a local signed build before we take it to CI.

ryanclark committed:
Remove unneeded code added from strict null checks

ryanclark commented on web/packages/teleterm/src/services/tshd/createClient.ts:
```diff
@@ -110,63 +119,64 @@ export function createTshdClient(
             reject(err);
           } else {
-            resolve(response.toObject().gatewaysList as types.Gateway[]);
+            resolve(response!.gateways as types.Gateway[]);
```
Yep, I've removed them

ravicious committed:
Move out dir to build/app

On masted it's build/app/dist.

We don't want to use build directly, as the app is being packaged in
build/release.

ravicious committed:
Move Node-only deps to dependencies

https://electron-vite.org/guide/dev#dependencies-vs-devdependencies

cross-env has been removed since it's not needed at all.

gzdunek commented on web/packages/teleterm/electron.vite.config.mts:
```diff
@@ -0,0 +1,169 @@
+
+function onwarn(warning, warn) {
+  // ignore Vite complaining about protobufs using eval
+  if (warning.code === 'EVAL') {
```
I don't see any warning related to `eval` if I remove this. I guess we don't need it.

ryanclark replied:
Ah, must've been from the other TypeScript protobuf gen library I had tried. I'll remove!

ryanclark committed:
Remove eval warning suppression

ravicious committed:
Fix default path of tsh

ravicious committed:
Rename dev-term to start-term-skip-native-deps

ravicious commented on package.json:
```diff
@@ -10,6 +10,7 @@
     "start-teleport-e": "yarn workspace @gravitational/teleport.e start",
     "build-term": "yarn workspace @gravitational/teleterm build",
     "start-term": "yarn workspace @gravitational/teleterm start",
+    "dev-term": "yarn workspace @gravitational/teleterm dev",
```
Initially we added native deps to start-term after the hackathon because Krzysztof and Edoardo had a ton of problems. They removed the whole node_modules each time they ran into some kind of a problem with JS and then they didn't remember about reinstalling native deps. ;)

So we just put native deps there. Grzegorz knew that it'll be slow as hell on Windows, but I was like "We'll deal with that later".

Well, last week I had to finally build the app on Windows and it was slow as hell. So it's good that you added it there because I'd probably end up doing the same in a week or two.

gzdunek committed:
Use correct `oneofKind` value for `ptyEventOneOfIsStartError`

gzdunek commented on web/packages/teleterm/src/ui/services/workspacesService/workspacesService.ts:
Most of the changes here can be reverted here, right? I think the are leftovers from strict null checks.

gzdunek commented on web/packages/teleterm/src/helpers.ts:
```diff
@@ -0,0 +1,200 @@
+  PtyServerEvent,
+} from 'teleterm/sharedProcess/api/protogen/ptyHostService_pb';
+
+export const ExcludesFalse = Boolean as any as <T>(x: T | false) => x is T;
```
This can be removed (altogether with the comment from line 19).

ryanclark committed:
Undo strict null check changes

ravicious commented:
Grzegorz and I want to go through Connect's test plan again once we merge this into master and make a dev build off of v15 with these changes. But what about the changes in Web UI, how are we going to test them?

ravicious commented:
If we merge it to master and backport to v15 only after a few days, it'll be inconvenient to anyone working on the Web UI code that interacts with types from protos. Once we get everything working in this PR, we should probably do tests on the v15 dev build and only then merge to master, followed by a quick backport to v15.

gzdunek committed:
Correctly construct env vars struct for shell

Based on https://stackoverflow.com/a/66989401

ravicious commented on web/packages/teleterm/package.json:
```diff
@@ -2,23 +2,21 @@
-    "start": "yarn build-native-deps && webpack serve --config webpack.renderer.dev.config.js --progress",
-    "start-main": "webpack build --config webpack.main.config.js --mode=development --progress --watch",
+    "start": "yarn build-native-deps && yarn start-skip-native-deps",
+    "start-skip-native-deps": "electron-vite dev",
```
Regarding hot reloading of main process and preload: I just tried it and it seems to work okay. You just do `yarn start-term -w` and electron-vite automatically restarts the app whenever the main process or the preload code changes. It restarts on changes to the shared process and the agent cleanup daemon too.

Restarting the app steals focus, that's one minus.

@gzdunek Can you think of a situation where we'd want to rebuild the main process but not restart the app immediately?

I feel like the `-w` option works well enough, we could work with it for a week or two and then see what could be adjusted. Besides, who cares about hot reloading when you can just rerun `yarn start-term` in a few seconds.

gzdunek replied:
>@gzdunek Can you think of a situation where we'd want to rebuild the main process but not restart the app immediately?

Nothing comes to my mind.

>I feel like the -w option works well enough, we could work with it for a week or two and then see what could be adjusted. Besides, who cares about hot reloading when you can just rerun yarn start-term in a few seconds.

I agree.

gzdunek replied:
I see that reloading doesn't work for the renderer:
```
MAIN [Main] warn: Navigation to https://localhost:8080/ blocked by 'will-navigate'
```
I will add an exception in dev mode for localhost:8080.

ravicious replied:
The previous mechanism we had with F6 restarting the app was useful in situations where you modified just tsh and wanted to restart the app. There's probably a way to set up a rollup watcher in electron-vite that's going to restart the app on changes to `build/tsh`, but that's not essential for now.

gzdunek committed:
Allow reloading renderer app

ravicious commented on .eslintignore:
```diff
@@ -3,3 +3,5 @@
 # Ignore WASM generated files:
 web/packages/teleport/src/ironrdp/pkg
+
+build
```
Did you mean to ignore `web/packages/teleterm/build`?

ravicious replied:
Eslint ignored the Webpack builds probably because of this:

https://github.com/gravitational/teleport/blob/218d608f835b8b15c2e8dca2a3542652d470ed4c/web/packages/build/.eslintrc.js#L37

And the files prepared by Webpack were in `web/packages/teleterm/build/app/dist`.

ravicious commented on web/packages/teleport/src/User/UserContext.tsx:
```diff
@@ -95,9 +93,12 @@ export function UserContextProvider(props: PropsWithChildren<unknown>) {
     }
-    clusterPreferences.current[clusterId].pinnedResources = pinnedResources;
+    clusterPreferences.current[clusterId].pinnedResources.resourceIds =
+      pinnedResources;
```
Could someone double check this change here?

ravicious commented on web/packages/teleterm/src/services/tshd/types.ts:
```diff
@@ -107,43 +105,33 @@ export interface Gateway extends apiGateway.Gateway.AsObject {
->;
+  apiService.GetRequestableRolesResponse;
+
+export type ReportUsageEventRequest = apiUsageEvents.ReportUsageEventRequest;
```
This change should be fine I think, @gzdunek AFAIK it was to accommodate for some weird Timestamp shenanigans in protos?

gzdunek replied:
I think so, I will make sure that the correct date is sent with the usage events.

gzdunek replied:
Except for the wrong environment, usage events are sent correctly.

ravicious committed:
Move JSDoc for TshdEventContextBridgeService to correct position

ravicious committed:
Remove outdated comment about type casting in tshdEvents/index.ts

ravicious commented on web/packages/teleterm/src/types.ts:
```diff
@@ -77,14 +79,21 @@ export type {
+
+export type ExtractResponseType<T> = T extends sendUnaryData<infer Res>
+  ? Res
+  : never;
```
This is so much nicer than how I implemented it!

ravicious committed:
Drop type cast for error.cause['message']

ravicious committed:
Ignore teleterm's build dir in .eslintignore

**/dist/** has been removed from ignorePatterns because it's no longer
necessary, neither Web UI nor Connect build in a dir matching this pattern.

ravicious reviewed (commented):
I went through the whole PR and everything appears to be okay just from looking at the code alone. We'll run the Connect test plan on 15.0.0-dev.ravicious.10 just to double check.

ravicious committed:
Define NODE_ENV for main and renderer

ravicious committed:
Define only NODE_ENV, not whole process.env

Otherwise process.env in the main process is always replaced with an
object containing just NODE_ENV.

We need process.env to be properly populated since it's then used when
spawning tsh daemon and other processes.

Using define like this guarantees that only callsites using
process.env.NODE_ENV directly will be affected.

ravicious reviewed (approved):
Okay, we ran the test plan against v15.0.0-dev.ravicious.10 and it looks everything works fine. What's next, do we wait for the protobuf PRs to land in master?

ryanclark commented:
> Okay, we ran the test plan against v15.0.0-dev.ravicious.10 and it looks everything works fine. What's next, do we wait for the protobuf PRs to land in master?

Yep

ryanclark committed:
Add support for the cluster preferences backend mismatch from the gRPC UserPreferences object

gzdunek commented on web/packages/teleport/src/services/userPreferences/userPreferences.ts:
```diff
@@ -108,3 +131,36 @@ export function makeDefaultUserClusterPreferences(): ClusterUserPreferences {
+): BackendUserPreferences {
+  return {
+    ...preferences,
+    clusterPreferences: {
+      pinnedResources:
+        preferences.clusterPreferences?.pinnedResources?.resourceIds ?? [],
+    },
```
I'd spread `clusterPreferences` too:
```
    clusterPreferences: {
      ...preferences.clusterPreferences,
      pinnedResources:
        preferences.clusterPreferences?.pinnedResources?.resourceIds ?? [],
    }
```
If someone adds a new property, we won't skip it. The same for `convertBackendUserPreferences`.

ryanclark committed:
Return resourceIds from preferences, convert preferences before updating

ryanclark committed:
Future proof all of the clusterPreferences object getting converted correctly

gzdunek commented:
Something is still wrong with leaf cluster preferences, I get JS errors. Let me look into it.

gzdunek committed:
Fix cluster preferences for leaf clusters

gzdunek committed:
Use correct `AssistViewMode` enum, remove the old one and unused interface

gzdunek committed:
Lint fix

ryanclark committed:
Add back in JS protobuf stuff, it can be removed in a later PR

ryanclark committed:
Remove `List` from property

ryanclark committed:
Add license header

gzdunek commented on web/packages/teleterm/src/ui/AppInitializer/AppInitializer.tsx:
```diff
@@ -35,12 +35,13 @@ export const AppInitializer = () => {
 
   const initializeApp = useCallback(async () => {
     try {
+      console.log('here1');
```
`console.log`s to remove.

ryanclark committed:
Remove console.logs
